bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java

```diff
@@ -55,119 +55,122 @@ private void initializeUIComponents() {
 		Notification.show("Saved successfully", Type.TRAY_NOTIFICATION);
 	}
-	
+
 	private void revertChanges() {
 		model.resetReportForEdit();
 		binder.setBean(model.getReportForEdit());
 	}
 
 	private void saveComment() {
-		if (!txtComment.getOptionalValue().isPresent())
-			return;
-		model.saveComment(txtComment.getValue(), BaseModel.loginUser);
+		if (!txtComment.isEmpty())
+			model.saveComment(txtComment.getValue(), BaseModel.loginUser);
+		model.saveAttachments();
 		fillComments();
-		txtComment.clear();
+		cleanAttachments();
 	}
 
-
 	@Override
 	public void uploadStarted(StartedEvent event) {
 		pnlAttachments.setVisible(true);
 		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
 		ProgressBar progress = new ProgressBar();
 		progress.setCaption(event.getFilename());
-		model.getAttachmentUIElements().put(event.getFilename(), progress);
+		model.getUploadingUIElements().put(event.getFilename(), progress);
 		layout.addComponent(progress);
 	}
 
-
 	@Override
 	public void uploadSucceeded(SucceededEvent event) {
-		ProgressBar progress = (ProgressBar) model.getAttachmentUIElements().get(event.getFilename());
+		ProgressBar progress = (ProgressBar) model.getUploadingUIElements().remove(event.getFilename());
 		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
 		layout.removeComponent(progress);
-		
 		FileResource file = new FileResource(new File(ReportsModel.FILEUPLOAD_PATH + event.getFilename()));
-		Link link = new Link(event.getFilename(), file);
-		link.setStyleName("tiny");
-		layout.addComponent(link);
-		model.getAttachmentUIElements().put(event.getFilename(), link);
 		try {
-			model.saveAttachment(event.getFilename(), event.getMIMEType(), file.getStream());
+			model.attachFile(event.getFilename(), event.getMIMEType(), file.getStream());
 		} catch (IOException e) {
 			Notification.show("Error occurred while uploading the file", e.getMessage(), Type.ERROR_MESSAGE);
 		}
-		//Refresh comments after new comment;
-		fillComments();
+		FileDownloader downloader = new FileDownloader(file);
+		Link link = new Link();
+		link.setCaption(event.getFilename());
+		link.setStyleName("tiny");
+		downloader.extend(link);
+		layout.addComponent(link);
+		commentsUpdated();
 	}
-	
 
 	@Override
 	public OutputStream receiveUpload(String filename, String mimeType) {
 		FileOutputStream fos = null;
 		try {
 			fos = new FileOutputStream(ReportsModel.FILEUPLOAD_PATH + filename);
 		} catch (FileNotFoundException e) {
-			e.printStackTrace();
+			Notification.show("Error occurred while uploading the file", e.getMessage(), Type.ERROR_MESSAGE);
 		} finally {
 			return fos;
 		}
 	}
 
-
 	@Override
 	public void updateProgress(long readBytes, long contentLength) {
 		ProgressBar bar = null;
-		for (Object obj : model.getAttachmentUIElements().values()) {
-			if (obj instanceof ProgressBar)
-				bar = (ProgressBar)obj;
+		for (Object obj : model.getUploadingUIElements().values()) {
```

Comment:
I didn't understand what this for is about..

Reply:
changes made for single upload